### ISSUE TITLE
Reimplement access control for changing which societies own a show

### ISSUE THREAD

**Issue by GKFX** [camdram/camdram#538]
From #536:

hoyes • about 2 hours ago
This bit was there to try and prevent the following:

Person who happens to be an admin of society X is putting on an independent show
Person initially lists society X as funding the show so that their urgent audition advert shows up immediately
Person removes Society X from the show immediately afterwards, and the show stays authorised.
This might be an over-compensation in hindsight, but maybe worth creating a ticket to discuss if no quick solution.
 
GKFX • 1 minute ago
Ah, interesting. The reason I removed that behaviour is partly because I couldn’t see the logical extension to the new code, and partly because as in the commit message it seems a bit bureaucratic.

Based on what you’ve said though, I guess the correct interpretation is: you may not take the show from a state you could authorize to one that you couldn’t. I’ll pull that out into a new ticket as it seems non-essential for this to go through.

**Comment by hoyes** [comment]
This seems like an annoyingly complicated problem, which goes to the root of exactly what is an authoriser authorising? E.g.:
 * What happens if someone is both an admin of the society and the venue for a show, and either the society and/or venue is removed?
 * If a co-funding society is added to a show at a later date, should that society maybe get a chance to authorise that change if it has active admins?
 * What happens if the venue admin who originally authorised your show is no longer a venue admin when you remove a society? Or vice versa (e.g. if the venue changes the other side of a society AGM...)

The most complete solution I can think of is:
 * Add a list of all the organisation_ids of the authoriser (at the moment that they authorise the show) to the show entry.
 * If any changes are made to the societies and/or venues of a show, compare the new set of organsiation_ids to the set of organsiation_ids above and ensure there is still a intersection.

...but that seems like a lot of work to support a relatively minor edge case.

The previous remedy (essentially forbidding non-admins from changing societies post-authorisation) kinda worked and I don't think it caused too much support churn. There might be some other solutions too.. e.g. if we did more work on #198 and made the list of changes that have been made to a show page public.

**Comment by hoyes** [comment]
I made some slightly lazy changes to:
 * The message that is displayed to a non-admin on a show page while it is awaiting approval
 * The contents on the email sent out to society owners when a show's societies are changed.

in c5074448e3b60a after multiple society changes.

Dumping this here as we it kinda relates to the wider issue of exactly who needs to be notified when regarding show authorisation.

**Comment by hoyes** [comment]
Also (from Gitter): Maybe a good idea to suppress email notifications when a society/venue change is made by a Camdram admin, to avoid spamming society/venue admins while doing batch processing.

**Comment by CHTJonas** [comment]
(sorry didn't see☝️in time before I created #546! Will leave though as I think it has slightly wider use-cases)

**Comment by CHTJonas** [comment]
Yeah this does seem a bit overcomplicated to me. To be honest I think I agree with what @hoyes suggests: show approval means that the basic info of venue, society (and maybe show name?) is locked-down. And then people can open support tickets if they need to. Either that or we just ignore this loophole and assume good faith?
